File: benchmarking/SquareRatio.java
package benchmarking;

/**
 * Square Ratio Finder
 * 
 * Given a ratio, finds two squares which divide
 * to that ratio.
 */


public class SquareRatio {

	public static void main(String[] args) {
		runTest(0.1241342, (int) 10e5);

		runTest(0.63, (int) 10e5);
		runTest(0.63, (int) 10e6);
		runTest(0.63, (int) 10e7);
	}

	static void runTest (double target, int iters) {
		double epsilon = 1e-10;

		int numer = 1;
		int denom = 1;

		double calculated = 0;

		long startTime = System.currentTimeMillis();

		for (int i=0; i<iters; i++) {
			calculated = (double) numer * numer / denom / denom;

			if (calculated - epsilon < target)
				numer++;
			else if (calculated + epsilon > target)
				denom++;
			else
				break;
		}

		long endTime = System.currentTimeMillis();

		System.out.println(String.format(
			"%d iterations, got %.4f = %d / %d",
			iters, calculated, numer, denom
		));

		System.out.println(String.format("(%d / %d)^2", numer, denom));
		System.out.println(String.format("%f seconds", (endTime - startTime) / 1000.0));
		System.out.println();
	}

}

File: benchmarking/SquareRatio.py
'''
Square Ratio Target Finder

This calculator with find an aproximation of a decimal
as the ratio of two squares.
'''

import time

def run_test (target, iters):
	iters = int(iters)

	eps = 10e-10

	numer = 1
	denom = 1

	calculated = 0

	start_time = time.time()

	for i in range(iters):
		calculated = numer ** 2 / denom ** 2
		# print("%.4f = %d / %d" % (calculated, numer, denom))

		if calculated - eps < target:
			numer += 1
		elif calculated + eps > target:
			denom += 1
		else:
			break

	end_time = time.time()

	print("%4d iterations, got %.4f = %d / %d, difference of %f" % (iters, calculated, numer**2, denom**2, abs(calculated - target)))
	print("(%d / %d)^2" % (numer, denom))
	print("%f seconds" % (end_time - start_time))
	print()



run_test(0.1241342, 10e5) # 0.665 sec

run_test(0.63, 10e5)      # 0.780 sec
run_test(0.63, 10e6)      # 6.577 sec
run_test(0.63, 10e7)      # 67.17 sec


